### ISSUE TITLE
Suboptimal performance with NDA augmentation in snn_train.py

### ISSUE THREAD

**Issue by modernwild** [myuansun/EventRPG#1]
# Description
When training with the default hyperparameters in snn_train.py using augment_mode="Identity", I obtained good results. However, when switching to augment_mode="NDA", the performance drops to ~88.5% after 200 epochs（expected result is 93.67%). Could you advise if there are specific hyperparameter configurations recommended for NDA/RPG augmentations?

**Comment by myuansun** [comment]
Hi @modernwild ,

Maybe you could try a weight decay of 5e-4 on the optimizer? All other hyperparameters we leverage are listed in the appendix of our paper.

**Comment by modernwild** [comment]
Thank you for your advice. I've adjusted weight_decay value but still observe gap in DVSGesture performance when using different augmentation methods with SEWResNet-18. Here are my training settings and results:  
● Identity: 94.4% (matches reported results)  
```shell
Namespace(l_mags=7, train_num_workers=16, train_batch_size=20, train_num_epochs=200, lr=0.0005, train_distributed=False, augment_mode='Identity', weight_decay=0, classifier='resnet18', timesteps=16, seed=0, use_wandb=False, relcam_mask_std_ratio=1, relcam_coord_std_ratio=0, spiking_neuron='PLIF', dataset='DVSGesture', relprop_mode='slrp', relevance_mix='layer4', mask='post', mix_prob=0.5, event_resolution=(128, 128), crop_dimension=(128, 128), num_classes=11, data_shape_prefix='TB')
```

● NDA: 89.0%
```shell
Namespace(l_mags=7, train_num_workers=16, train_batch_size=20, train_num_epochs=200, lr=0.0005, train_distributed=False, augment_mode='NDA', weight_decay=0.0005, classifier='resnet18', timesteps=16, seed=0, use_wandb=False, relcam_mask_std_ratio=1, relcam_coord_std_ratio=0, spiking_neuron='PLIF', dataset='DVSGesture', relprop_mode='slrp', relevance_mix='layer4', mask='post', mix_prob=0.5, event_resolution=(128, 128), crop_dimension=(128, 128), num_classes=11, data_shape_prefix='TB')
```

● RPG: 91.0% 
```shell
Namespace(l_mags=7, train_num_workers=16, train_batch_size=20, train_num_epochs=200, lr=0.0005, train_distributed=False, augment_mode='RPG', weight_decay=0.0005, classifier='resnet18', timesteps=16, seed=0, use_wandb=False, relcam_mask_std_ratio=1, relcam_coord_std_ratio=0, spiking_neuron='PLIF', dataset='DVSGesture', relprop_mode='slrp', relevance_mix='layer4', mask='post', mix_prob=0.5, event_resolution=(128, 128), crop_dimension=(128, 128), num_classes=11, data_shape_prefix='TB')
```

I noticed something odd, NDA and RPG results are way off from what's in the paper. Do I need to tweak batch sizes or learning rates between different augmentation methods for DVSGesture? Or am I missing some hidden details here?

**Comment by myuansun** [comment]
Hi @modernwild ,

I carefully compared the GitHub version with the original implementation and did find one issue.

Please replace lines 146-156 of `snn_train.py` with:

```
aug_this_iter = True if random.random() < 0.5 else False
if aug_this_iter:
    if args.augment_mode == 'RPG':
        aug_events, labels, spatial_mixup = event_augment.batch_augment(events, labels, args.mask,
                                                                        mix_prob=args.mix_prob)
    elif args.augment_mode == 'NDA':
        aug_events = event_augment.nda_batch_augment(events)
        if random.random() < 0.5:
            spatial_mixup = True
            aug_events, labels = event_augment.cut_mix(aug_events, labels)
    elif args.augment_mode == 'eventdrop':
        aug_events = event_augment.eventdrop_batch_augment(events)
    else:
        aug_events = events
else:
    aug_events = events
```

This means reducing the probability of data augmentation from 1 to 0.5, which I implemented in the DVSGesture training code.

**Comment by modernwild** [comment]
Thank you @myuansun ,

I've reduced the probability to 0.5 and rerun the experiments. The results show ~94% (NDA) and ~94.2% (RPG) accuracy - all other training settings remained the same. From these results, it seems the data augmentation doesn't work. 

Could you please check if there might be any implementation issues in `utils/augment.py` ?

**Comment by myuansun** [comment]
Hi @modernwild ,

After comparison, augment.py does not differ significantly from the GitHub version. However there are some parameters that could influence the performance, including the `alpha` of `class RelavanceCAM` (might be 2) and `relcam_coord_std_ratio` (might be 0.5). By the way, switching to SLTRP (set `relprop_mode` to `sltrp`) and Saliency Map (set `relevance_mix` to `long`) usually yields a better result.

**Comment by modernwild** [comment]
It works！Thank you for your help @myuansun. By the way, should the probability of augmentation also be changed to 0.5 for other datasets?

**Comment by myuansun** [comment]
Hi @modernwild ,

It's good to see it works. I'm not really sure about the specific probability for every dataset, but I'm sure it is either 1 or 0.5. From my perspective and experience, a lower probability would be better for SNN training and a higher probability is more suitable for ANN training.
